allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -335,5 +289,122 @@ def deserialize_instance(self, string: str) -> Instance:
         The default implementation is to use `jsonpickle`.  If you would like some other format
         for your pre-processed data, override this method.
         """
+        return jsonpickle.loads(string.strip())  # type: ignore
 
-        return jsonpickle.loads(string)  # type: ignore
+    def _multi_worker_islice(
+        self, iterable: Iterable[Any], transform: Optional[Callable[[Any], Instance]] = None,
+    ) -> Iterable[Instance]:
+        """
+        Helper method that determines which raw instances to skip based on the current
+        node rank (for distributed training) and worker ID (for multi-process data loading).
+
+        # Parameters
+
+        iterable : `Iterable[Any]`
+            An iterable that yields raw data that can be transformed into `Instance`s
+            through the `transform` function.
+        transform : `Optional[Callable[[Any], Instance]]`, optional (default = `None`)
+            An optional function that will be applied to the raw data generated
+            by `iterable` to create `Instance`s.
+
+        # Returns
+
+        `Iterable[Instance]`
+        """
+        start_index = 0
+        step_size = 1
+        if not self.manual_distributed_sharding and util.is_distributed():
+            start_index = dist.get_rank()
+            step_size = dist.get_world_size()
+        worker_info = None if self.manual_multi_process_sharding else get_worker_info()
+        if worker_info:
+            # Scale `start_index` by `num_workers`, then shift by worker `id`.
+            start_index *= worker_info.num_workers
+            start_index += worker_info.id
+            # Scale `step_size` by `num_workers`.
+            step_size *= worker_info.num_workers
+
+        islice = itertools.islice(iterable, start_index, self.max_instances, step_size)
+        if transform is not None:
+            return (transform(x) for x in islice)
+        return islice
+
+    def _instance_iterator(self, file_path: str) -> Iterable[Instance]:
```

Comment:
Yea, there's a lot going on here.

Most of this comes from the now removes `_LazyInstances` class: https://github.com/allenai/allennlp/blob/master/allennlp/data/dataset_readers/dataset_reader.py#L72-L101

The additional complexity is from managing a file lock which is necessary to avoid bugs that might be very hard to catch.